### PR TITLE
Add `AutoDiscoveryService` to run auto discovery steps in parallel

### PR DESCRIPTION
**Pull request by cketti**
I'll change the result type in a follow-up PR.

### CODE REVIEW

File: feature/autodiscovery/service/src/main/kotlin/app/k9mail/autodiscovery/service/PriorityParallelRunner.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * Runs a list of [AutoDiscoveryRunnable]s with descending priority in parallel and returns the result with the highest
+ * priority.
+ *
+ * As soon as an [AutoDiscoveryRunnable] returns a non-`null` result, runnables with a lower priority are canceled.
+ */
+internal class PriorityParallelRunner(
+    private val runnables: List<AutoDiscoveryRunnable>,
+    private val coroutineDispatcher: CoroutineDispatcher = Dispatchers.IO,
+) {
+    suspend fun run(): AutoDiscoveryResult? {
```
Comment (wmontwe) [inline feature/autodiscovery/service/src/main/kotlin/app/k9mail/autodiscovery/service/PriorityParallelRunner.kt:23]: Here is an updated version with some changes:

```kotlin
    private val mutex = Mutex()

    suspend fun run(): AutoDiscoveryResult? {
        return coroutineScope {
            val deferredList = mutableListOf<Deferred<AutoDiscoveryResult?>>()

            suspend fun cancelLowerPriority(index: Int) {
                if (index == deferredList.lastIndex) return

                mutex.withLock {
                    for (deferred in deferredList.subList(fromIndex = index + 1, toIndex = deferredList.size)) {
                        deferred.cancel()
                    }
                }
            }

            runnables.mapIndexed { index, runnable ->
                async(coroutineDispatcher, start = LAZY) {
                    runnable.run()?.also {
                        // We've got a non-null result, so cancel all coroutines with lower priority.
                        cancelLowerPriority(index)
                    }
                }.also { deferredList.add(it) }
            }

            deferredList.forEach { it.start() }

            @Suppress("SwallowedException")
            deferredList.firstNotNullOfOrNull { deferred ->
                try {
                    deferred.await()
                } catch (e: CancellationException) {
                    null
                }
            }
        }
    }
```

**Review by wmontwe — COMMENTED** [review]
I had a look and maybe some ideas how to improve the `PriorityParallelRunner`

File: feature/autodiscovery/service/src/main/kotlin/app/k9mail/autodiscovery/service/PriorityParallelRunner.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        return coroutineScope {
+            var deferredList = listOf<Deferred<AutoDiscoveryResult?>>()
+
+            fun cancelLowerPriority(index: Int) {
+                if (index == deferredList.lastIndex) return
+
+                for (deferred in deferredList.subList(fromIndex = index + 1, toIndex = deferredList.size)) {
```
[the remark this answers is not in the document]
Comment (cketti) [reply feature/autodiscovery/service/src/main/kotlin/app/k9mail/autodiscovery/service/PriorityParallelRunner.kt:30]: The list is not modified by any of the threads. Concurrently reading it from multiple threads should be fine.

[the remark this answers is not in the document]
Comment (wmontwe) [reply feature/autodiscovery/service/src/main/kotlin/app/k9mail/autodiscovery/service/PriorityParallelRunner.kt:30]: Right, the list is sequentialy updated and not by the executed coroutines. It's not needed.

**Review by wmontwe — APPROVED** [review]
LGTM

**Pull request merged**